Rename std::fs::soft_link to std::fs::symlink

lambda opened the pull request:
Implement [RFC #1048][rfc].

On Windows, when you create a symbolic link you must specify whether it
points to a directory or a file, even if it is created dangling, while
on Unix, the same symbolic link could point to a directory, a file, or
nothing at all.  Furthermore, on Windows special privilege is necessary
to use a symbolic link, while on Unix, you can generally create a
symbolic link in any directory you have write privileges to.

This means that it is unlikely to be able to use symbolic links purely
portably; anyone who uses them will need to think about the cross
platform implications.  This means that using platform-specific APIs
will make it easier to see where code will need to differ between the
platforms, rather than trying to provide some kind of compatibility
wrapper.

Furthermore, `soft_link` has no precedence in any other API, so to avoid
confusion, move back to the more standard `symlink` terminology.

Create a `std::os::unix::symlink` for the Unix version that is
destination type agnostic, as well as `std::os::windows::{symlink_file,
symlink_dir}` for Windows.

Because this is a stable API, leave a compatibility wrapper in
`std::fs::soft_link`, which calls `symlink` on Unix and `symlink_file`
on Windows, preserving the existing behavior of `soft_link`.

[rfc]: https://github.com/rust-lang/rfcs/pull/1048

rust-highfive commented:
r? @huonw

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #24433) made this pull request unmergeable. Please resolve the merge conflicts.

lambda commented:
I've rebased and updated to include the split into `std::os::unix::symlink` and `std::os::windows::{symlink_file, symlink_dir}` per the latest discussion on rust-lang/rfcs#1048.

I don't have Windows available to build or test on, and tests for symlinks on a buildbot don't work properly as you need special permissions to create symlinks on Windows, so if it would be possible for someone to try it out and make sure that I didn't break anything on Windows I would appreciate that.

r? @alexcrichton

alexcrichton commented on src/libstd/sys/unix/ext.rs:
````diff
@@ -220,6 +224,27 @@ pub mod fs {
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn symlink_file<P: AsRef<Path>, Q: AsRef<Path>>
````
symlink_file => symlink ?

lambda replied:
Whoops, fixed.

alexcrichton commented on src/libstd/sys/windows/fs2.rs:
```diff
@@ -402,11 +402,16 @@ pub fn readlink(p: &Path) -> io::Result<PathBuf> {
     use sys::c::compat::kernel32::CreateSymbolicLinkW;
     let src = to_utf16(src);
     let dst = to_utf16(dst);
+    let flags = if dir { 1 } else { 0 };
```
Could you add this constant of `1` to the `sys::c` module with the name `SYMBOLIC_LINK_FLAG_DIRECTORY` as well?

lambda replied:
Done.

alexcrichton commented on src/libstd/sys/unix/ext.rs:
```diff
@@ -220,6 +224,27 @@ pub mod fs {
+
+    /// Creates a new symbolic link on the filesystem.
+    ///
+    /// The `dst` path will be a symbolic link pointing to the `src` path.
```
These docs are the ones that primarily show up (because our doc builders are linux), so could you expand this a bit to include the information about Windows as well? Just something along the lines of "If you're on windows...".

lambda replied:
Done.

alexcrichton replied:
Could you also add some notes about how windows requires a special permission to create symlinks? Otherwise this all looks great to me!

lambda replied:
Sure, added a brief note to that effect. I was a little hard to summarize the situation without a very verbose explanation, so I just made sure to include the name of the specific privilege required that allows you to find the explanation on StackOverflow.

alexcrichton replied:
Looks good to me, thanks!

alexcrichton commented:
Looks great, thanks @lambda! It's ok to not worry about symlink tests for windows right now.

alexcrichton commented on src/libstd/fs.rs:
````diff
@@ -828,17 +832,20 @@ pub fn hard_link<P: AsRef<Path>, Q: AsRef<Path>>(src: P, dst: Q) -> io::Result<(
 /// ```
+#[deprecated(since = "1.0.0",
+             reason = "replaced with std::os::unix::fs::symlink and \
+                       std::os::windows::{symlink_file, symlink_dir}")]
````
Oh oops, `windows::{` -> `windows::fs::{`

lambda replied:
Whoops, thanks for the catch.

alexcrichton commented on src/libstd/sys/unix/ext.rs:
````diff
@@ -220,6 +224,36 @@ pub mod fs {
+    /// # }
+    /// ```
+    pub fn symlink<P: AsRef<Path>, Q: AsRef<Path>>
+        (src: P, dst: Q) -> io::Result<()>
````
Stylistically here we tend to do:

```rust
pub fn symlink<P: AsRef<Path>, Q: AsRef<Path>>(src: P, dst: Q)
                                               -> io::Result<()>

lambda replied:
Sure, rewrapped; I wasn't quite sure what the convention should be. Turns out this one fits all on one line anyhow, but wrapped the Windows ones this way.

alexcrichton commented:
I've now merged the RFC, just noticed a few nits though, but other than that r=me

lambda commented:
Fixed up the nits.

r? @alexcrichton

alexcrichton commented:
@bors: r+ acd426bda5166eee69d864c62b499a6217ca3309

Thanks!

bors commented:
:hourglass: Testing commit acd426b with merge caa75f0...

bors commented:
:broken_heart: Test failed - [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/3052)

lambda commented:
Added missing `use` statement.

r? @alexcrichton

alexcrichton commented:
@bors: r+ 14044c9

bors commented:
:hourglass: Testing commit 14044c9 with merge 1556209...

bors commented:
:broken_heart: Test failed - [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/4537)

lambda commented:
Added feature gates to doc tests, and actually ran doc tests this time to ensure they work.

r? @alexcrichton

alexcrichton commented:
@bors: r+ 1489e23

bors commented:
:hourglass: Testing commit 1489e23 with merge feb8acd...

bors commented:
:broken_heart: Test failed - [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/4574)

lambda commented:
:disappointed: Hmm, this failure doesn't look like it's related to my change; it looks like it just hung for some unknown reason.

alexcrichton commented:
@bors: retry

lambda commented:
I realized that the Windows doc tests were likely to fail, because I included examples in the Windows docs, but as discussed, you can't create symlinks without special privileges.

I've updated the Windows example to use `ignore` to prevent this. Apologies for the churn.

r? @alexcrichton

alexcrichton commented:
@bors: r+  3cc84ef

No worries!
